// nexmo-conversation/test/base/utils.js
// This file includes incoming events from the service
// for use in the unit tests. The default values are set here
// and should be modified in the test for changes.
// e.g. to set the id : utils.events_from_cs.member_joined_app.id = 3;

var events_from_cs = {
	member_joined_app: {
		"from": "MEM-47dbb905-6395-4ab8-89a9-db29d5eef3ac",
		"body": {
			"user": {
				"user_id": "USR-6d149825-4902-411c-b1d0-2bd3d91ea518",
				"name": "demo"
			},
			"timestamp": {
				"joined": "2017-12-08T09:35:08.522Z"
			}
		},
		"id": 1,
		"timestamp": "2017-12-08T09:35:08.523Z",
		"cid": "CON-aa4d0392-2317-4d6e-a160-dc580149755e",
		"type": "member:joined"
	},
	member_invited_app: {
		"from": "MEM-96b63e92-dcf3-4fc3-b3d8-59acd471ead6",
		"body": {
			"cname": "NAM-4dd5641a-c9ee-45d6-9a57-e0ed12488be6",
			"invited_by": "demo3",
			"user": {
				"member_id": "MEM-96b63e92-dcf3-4fc3-b3d8-59acd471ead6",
				"user_id": "USR-6d149825-4902-411c-b1d0-2bd3d91ea518",
				"media": {},
				"name": "demo"
			},
			"timestamp": {
				"invited": "2017-12-19T15:56:26.226Z"
			}
		},
		"id": 2,
		"timestamp": "2017-12-19T15:56:26.229Z",
		"cid": "CON-b0fe6af3-b831-4b15-bae7-fce13b9e758c",
		"type": "member:invited"
	},
	member_left_app: {
		"body": {
			"user": {
				"name": "demo",
				"id": "USR-6d149825-4902-411c-b1d0-2bd3d91ea518"
			},
			"timestamp": {
				"joined": "2017-12-22T11:47:40.485Z",
				"left": "2017-12-22T11:47:53.994Z"
			}
		},
		"from": "MEM-c26619a3-817f-40b0-86c3-9f0c68ec783e",
		"id": 2,
		"timestamp": "2017-12-22T11:47:53.994Z",
		"cid": "CON-895a7d17-97b5-449e-bc48-05c5e7e88bc1",
		"type": "member:left"
	},
	member_join_phone: {
		"from": "member_id",
		"body": {
			"user": {
				"id": "user_id"
			},
			"channel": {
				"type": "phone",
				"id": "channel id",
				"from": {
					"type": "phone",
					"number": "badger"
				},
				"to": {
					"type": "phone",
					"number": "111112456785"
				}
			},
			"timestamp": {
				"joined": "2017-11-17T17:24:25.725Z"
			}
		},
		"id": 10,
		"timestamp": "2017-11-17T17:24:25.738Z",
		"cid": "cid",
		"type": "member:joined"
	},
	conversation_join_success: {
		"rid": "506bf25a-d039-41de-afcd-63db85b1741d",
		"body": {
			"id": "MEM-04a9cef3-e3e2-4a6b-b8eb-66a7a6e83bde",
			"user_id": "USR-6d149825-4902-411c-b1d0-2bd3d91ea518",
			"state": "JOINED",
			"timestamp": {
				"joined": "2017-12-11T18:07:22.994Z"
			},
			"channel": {
				"type": "app"
			}
		},
		"type": "conversation:join:success",
		"cid": "CON-63a6feef-68a0-48f2-9bec-476e03ca9dfa"
	},
	conversation_get_success: {
		"name": "NAM-50949969-a277-4690-89c7-ec872c03e4df",
		"id": "CON-518e242f-d1b3-443d-99e0-9fd522c403f6",
		"timestamp": {
			"created": "2017-12-08T15:32:37.427Z"
		},
		"sequence_number": 2,
		"numbers": {

		},
		"properties": {

		},
		"members": [{
				"member_id": "MEM-570c99f4-dd7b-40e9-82c5-a5011d10e4e0",
				"user_id": "USR-6d149825-4902-411c-b1d0-2bd3d91ea518",
				"name": "demo",
				"state": "JOINED",
				"timestamp": {
					"joined": "2017-12-08T15:32:37.666Z"
				}
			},
			{
				"member_id": "MEM-dcee05a1-44ce-436c-9273-7eb502b22141",
				"user_id": "USR-0f3c52cb-e186-4aca-a52a-8e9fceb235d9",
				"name": "demo4",
				"state": "INVITED",
				"timestamp": {
					"invited": "2017-12-08T15:32:48.076Z"
				},
				"invited_by": "demo"
			}
		]
	}
}

module.exports = {
	events_from_cs: events_from_cs
};
