// nexmo-conversation/lib/member.js
/*
 * Conversation SDK
 *  Member object model
 *
 * Copyright (c) Nexmo Inc. 2015
 */

var WildEmitter = require('wildemitter');
var User = require('./user').User;
var Event = require('./events/event').Event;
var TextEvent = require('./events/text_event').TextEvent;
var ImageEvent = require('./events/image_event').ImageEvent;

// Conversation Client Error
var ConversationClientError = require('./conversationClientError');

var NexmoApiError = ConversationClientError.NexmoApiError;
var NexmoClientError = ConversationClientError.NexmoClientError;

/**
 * An individual user (i.e. conversation member).
 * @class
 * @param {Conversation} conversation
 * @param {Object} params 
 */
var Member = function(conversation, params) {
    this.conversation = conversation;

    //TODO: to move this to a constructor
    this._normalise = function(params) {
        if (params) {
            for (var key in params) {
                switch (key) {
                    case "member_id":
                        this.id = params.member_id;
                        break;
                    case "timestamp":
                        this.timestamp = params.timestamp;
                        break;
                    case "state":
                        this.state = params.state;
                        break;
                    default:
                        this[key] = params[key];
                }
            }

            if (params.from) {
                if (params.from) this.id = params.from;
            }
            this.user = this.user || {};
            if (params.user_id) {
                this.user.id = params.user_id;
            }
            if (params.name) {
                this.user.name = params.name;
            }
            if (params.user) {
                this.user = {
                    name: params.user.name,
                    id: params.user.user_id || params.user.id
                };
            }
            if (params.invited_by) this.invited_by = params.invited_by;
            this.channel = params.channel || {
                type: "app"
            };
            //join conversation returns our member with only id, 
            // compare it for now and use the username we have in the application object
            if (this.conversation.application.me && params.user_id === this.conversation.application.me.id) {
                this.user.name = this.conversation.application.me.name;
            }

            //make sure we don't keep a member.user_id, username in any flow
            delete this.user_id;
            delete this.name;
            delete this.user.user_id;
        }
    }
    // call WildEmitter constructor
    WildEmitter.call(this);
    this._normalise(params);
};

Member.prototype = new WildEmitter();
//reset the original Type
Member.prototype.constructor = Member;

/*
 * Play the given stream only to this member within the conversation
 * 
 * @param {String} [params] 
 * 
 * @returns {Promise<Event>} - not tested yet
 */
Member.prototype.playStream = function(params) {
    var self = this;
    var p = new Promise(function(resolve, reject) {
        self.conversation.application.session.sendRequest({
            type: 'audio:play',
            cid: self.id,
            to: self.id,
            body: params
        }, function(response) {
            if (response.type === 'audio:play:success') {
                var event = new Event(self.conversation, response);
                resolve(event);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

/** 
 * Speak the given text only to this member within the conversation
 * 
 * @param {String} [params] 
 * 
 * @returns {Promise<Event>}
 */
Member.prototype.sayText = function(params) {
    var self = this;
    var p = new Promise(function(resolve, reject) {
        self.conversation.application.session.sendRequest({
            type: 'audio:say',
            cid: self.id,
            from: self.conversation.me.id,
            to: self.id,
            body: params
        }, function(response) {
            if (response.type === 'audio:say:success') {
                var event = new Event(self.conversation, response);
                resolve(event);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

/**
 * Kick this member from the conversation
 * 
 * @returns {Promise}
 */
Member.prototype.kick = function() {
    var self = this;
    var p = new Promise(function(resolve, reject) {
        self.conversation.application.session.sendRequest({
            type: 'conversation:member:delete',
            cid: self.conversation.id,
            from: self.conversation.me.id,
            to: self.id,
            body: {
                "member_id": self.id
            }
        }, function(response) {
            if (response.type === 'conversation:member:delete:success') {
                resolve(response.body);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

/**
 * @ignore
 * Mute this member
 * 
 * @param {Boolean} [params] is muted
 * 
 * @returns {Promise} - 
 */
Member.prototype.mute = function(mute) {
    var self = this;
    var params = {};
    var p = new Promise(function(resolve, reject) {
        var type = 'audio:mute:off';
        if (mute) {
            type = 'audio:mute:on';
        }
        self.conversation.application.session.sendRequest({
            type: type,
            cid: self.id,
            to: self.id,
            body: params
        }, function(response) {
            if (response.type === 'audio:mute:success') {
                resolve(response.body);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

/*
 * Control the volume of this member
 * 
 * @param {String} [params] 
 * 
 * @returns {Promise} - not tested yet
 */
Member.prototype.volume = function(up) {
    var self = this;
    var params = {};
    var p = new Promise(function(resolve, reject) {
        var type = 'audio:volume:down';
        if (up) {
            type = 'audio:volume:up';
        }
        self.conversation.application.session.sendRequest({
            type: type,
            cid: self.id,
            to: self.id,
            body: params
        }, function(response) {
            if (response.type === 'audio:volume:success') {
                resolve(response.body);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

/**
 * @ignore
 * Earmuff this member
 * 
 * @param {Boolean} [params] 
 * 
 * @returns {Promise} 
 */
Member.prototype.earmuff = function(earmuff) {
    var self = this;
    var params = {};
    var p = new Promise(function(resolve, reject) {
        if (self.me === null) {
            reject(new NexmoClientError("error:self"));
        } else {
            var type = 'audio:earmuff:off';
            if (earmuff) {
                type = 'audio:earmuff:on';
            }
            self.conversation.application.session.sendRequest({
                type: type,
                cid: self.id,
                to: self.id,
                body: params
            }, function(response) {
                if (response.type === 'audio:earmuff:success') {
                    resolve(response.body);
                } else {
                    reject(new NexmoApiError(response));
                }
            });
        }
    });
    return p;
};

/*
 * Record this member
 * 
 * @param {String} [params] 
 * 
 * @returns {Promise} - not tested yet
 */
Member.prototype.record = function(params) {
    var self = this;
    var p = new Promise(function(resolve, reject) {
        self.conversation.application.session.sendRequest({
            type: 'audio:record',
            cid: self.id,
            to: self.id,
            body: params
        }, function(response) {
            if (response.type === 'audio:record:success') {
                var event = new Event(this.conversation, response);
                resolve(event);
            } else {
                reject(new NexmoApiError(response));
            }
        });
    });
    return p;
};

Member.prototype._handleEvent = function(event) {
    switch (event.type) {
        case 'member:invited':
            this._normalise(event.body);
            this.state = 'INVITED';
            this.timestamp.invited = event.body.timestamp.invited;
            break;
        case 'member:joined':
            this._normalise(event.body);
            this.state = 'JOINED';
            this.timestamp.joined = event.body.timestamp.joined;
            break;
        case 'member:left':
            this._normalise(event.body);
            this.state = 'LEFT';
            this.timestamp.left = event.body.timestamp.left;
            break;
        case 'member:media':
            this.media = event.body;
            break;
        default:
            break;
    }
};

module.exports = {
    Member: Member
};
